File: frontend/lib/privacy.ts
import type { Role } from "@/lib/types";

/**
 * Versi Kebijakan Privasi yang sedang berlaku.
 *
 * Wajib sama dengan PRIVACY_POLICY_VERSION di backend/core/privacy.py; tes
 * backend membaca berkas ini dan gagal bila keduanya berbeda. Mengubah versi
 * membuat setiap pengguna diminta menyetujui ulang sebelum melanjutkan, sesuai
 * Pasal 21 ayat (2) UU No. 27 Tahun 2022: perubahan informasi diberitahukan
 * sebelum terjadi.
 */
export const PRIVACY_POLICY_VERSION = "2026-09-25";
export const PRIVACY_POLICY_EFFECTIVE_LABEL = "25 September 2026";

/**
 * Surel untuk permohonan hak subjek data (Pasal 14: permohonan tercatat).
 * Diisi lewat lingkungan deploy, bukan ditanam di kode, supaya alamat pribadi
 * anggota tim tidak ikut terpublikasi tanpa keputusan tim.
 */
export const PRIVACY_CONTACT_EMAIL =
  process.env.NEXT_PUBLIC_PRIVACY_CONTACT_EMAIL?.trim() || null;

export type ConsentItemKey =
  | "data_perkuliahan"
  | "analisis_proses"
  | "usia_atau_izin_wali"
  | "tanggung_jawab_dosen"
  | "analisis_luar_negeri";

export interface ConsentItem {
  key: ConsentItemKey;
  title: string;
  body: string;
  required: boolean;
  /** Bagian Kebijakan Privasi yang menjelaskan butir ini lebih rinci. */
  anchor: string;
}

/**
 * Butir persetujuan per peran. Dipisah per tujuan karena Pasal 22 ayat (4)
 * mewajibkan tujuan yang berbeda dapat dibedakan secara jelas, dalam format
 * yang mudah dipahami dan bahasa yang sederhana. Kuncinya sama dengan
 * REQUIRED_ITEMS dan OPTIONAL_ITEMS di backend.
 */
export const CONSENT_ITEMS: Record<Role, ConsentItem[]> = {
  student: [
    {
      key: "data_perkuliahan",
      title: "Data akun dan perkuliahan",
      body: "ThinkPath menyimpan akunmu, kelas yang kamu ikuti, jawaban tugas, nilai, dan umpan balik dosen untuk menjalankan perkuliahan. Datanya disimpan dan diproses oleh penyedia cloud di luar Indonesia: Supabase untuk basis data (Singapura), serta Hugging Face dan Vercel untuk server aplikasi (perusahaan Amerika Serikat).",
      required: true,
      anchor: "data-yang-dikumpulkan",
    },
    {
      key: "analisis_proses",
      title: "Perekaman proses menulis dan analisis otomatis",
      body: "Selama form jawaban terbuka, ThinkPath mencatat jumlah kata setiap 30 detik, waktu mulai dan kumpul, serta berapa kali kamu merevisi, bukan isi ketikanmu. Jawabanmu lalu dianalisis otomatis untuk level Bloom dan indikasi penggunaan AI. Hasilnya bahan tinjau dosen pengampu kelasmu, bukan keputusan, dan tidak pernah memicu sanksi otomatis.",
      required: true,
      anchor: "analisis-otomatis",
    },
    {
      key: "usia_atau_izin_wali",
      title: "Usia",
      body: "Saya berusia 18 tahun atau lebih. Kalau belum, orang tua atau wali saya sudah menyetujui pemrosesan ini.",
      required: true,
      anchor: "anak",
    },
    {
      key: "analisis_luar_negeri",
      title: "Analisis tambahan oleh penyedia di luar negeri",
      body: "Izinkan teks jawabanmu dikirim ke Groq (Amerika Serikat) dan Winston AI (Kanada) untuk analisis yang lebih lengkap. Yang dikirim hanya teks jawaban, ditambah jenjang studi untuk Groq, tanpa nama atau email. Winston menyimpan teks yang dinilai sebagai laporan sampai laporan itu dihapus. Tanpa izin ini, jawabanmu tetap dianalisis, tetapi hanya dengan metode heuristik di server ThinkPath; hasilnya bisa kurang akurat, dan dosen melihat bahwa skornya berasal dari heuristik.",
      required: false,
      anchor: "penyedia-layanan",
    },
  ],
  teacher: [
    {
      key: "data_perkuliahan",
      title: "Data akun dan kelas",
      body: "ThinkPath menyimpan akun Anda, kelas, tugas, materi, dan penilaian yang Anda berikan. Datanya disimpan dan diproses oleh penyedia cloud di luar Indonesia: Supabase untuk basis data (Singapura), serta Hugging Face dan Vercel untuk server aplikasi (perusahaan Amerika Serikat).",
      required: true,
      anchor: "data-yang-dikumpulkan",
    },
    {
      key: "tanggung_jawab_dosen",
      title: "Tanggung jawab atas data mahasiswa",
      body: "Saya memakai data mahasiswa hanya untuk perkuliahan di kelas saya dan menjaga kerahasiaannya. Skor indikasi AI dan level Bloom adalah bahan tinjau dengan keterbatasan yang dijelaskan di Kebijakan Privasi. Saya tidak menjatuhkan sanksi akademik hanya berdasarkan skor, dan memberi mahasiswa kesempatan menjelaskan jawabannya, misalnya lewat sesi diskusi jawaban.",
      required: true,
      anchor: "analisis-otomatis",
    },
  ],
};

File: backend/core/privacy.py
"""Persetujuan pemrosesan data pribadi menurut UU No. 27 Tahun 2022 (UU PDP).

Satu versi Kebijakan Privasi berlaku pada satu waktu. Isinya ada di frontend
(`/kebijakan-privasi`), dan versinya harus sama dengan konstanta di bawah;
tes memeriksa keduanya supaya tidak menyimpang diam-diam.

Butir persetujuan dipisah per tujuan karena Pasal 22 ayat (4) mewajibkan
permintaan persetujuan yang memuat tujuan lain dapat dibedakan secara jelas.
Persetujuan yang tidak memenuhi ketentuan itu batal demi hukum (ayat (5)).
"""
from __future__ import annotations

from collections.abc import Iterable
from datetime import timedelta
from uuid import UUID

from django.utils import timezone

from .models import ConsentAction, ConsentRecord, Role

# Tanggal berlaku versi kebijakan. Mengubahnya membuat semua persetujuan lama
# tidak berlaku, sehingga setiap pengguna diminta membaca dan menyetujui ulang
# sebelum pemrosesan berlanjut dengan ketentuan baru (Pasal 21 ayat (2)).
PRIVACY_POLICY_VERSION = "2026-09-25"

# Data akun, kelas, jawaban, nilai, dan umpan balik untuk perkuliahan,
# termasuk penyimpanannya di penyedia cloud di luar Indonesia.
ITEM_COURSE_DATA = "data_perkuliahan"
# Perekaman jumlah kata tiap 30 detik dan analisis otomatis (level Bloom dan
# indikasi penggunaan AI) sebagai bahan tinjau dosen.
ITEM_PROCESS_ANALYSIS = "analisis_proses"
# Berusia 18 tahun ke atas, atau orang tua/wali sudah menyetujui (Pasal 25).
ITEM_AGE = "usia_atau_izin_wali"
# Dosen: memakai data mahasiswa hanya untuk kelasnya, menjaga kerahasiaan,
# dan tidak menjatuhkan sanksi hanya berdasarkan skor.
ITEM_TEACHER_DUTIES = "tanggung_jawab_dosen"
# Opsional: teks jawaban boleh dikirim ke Groq (AS) dan Winston AI (Kanada).
# Tanpa butir ini, jawaban hanya dianalisis heuristik di server ThinkPath.
ITEM_EXTERNAL_AI = "analisis_luar_negeri"

REQUIRED_ITEMS: dict[str, tuple[str, ...]] = {
    Role.STUDENT: (ITEM_COURSE_DATA, ITEM_PROCESS_ANALYSIS, ITEM_AGE),
    Role.TEACHER: (ITEM_COURSE_DATA, ITEM_TEACHER_DUTIES),
}
OPTIONAL_ITEMS: dict[str, tuple[str, ...]] = {
    Role.STUDENT: (ITEM_EXTERNAL_AI,),
    Role.TEACHER: (),
}


def latest_record(profile_id: UUID) -> ConsentRecord | None:
    return (
        ConsentRecord.objects.filter(profile_id=profile_id)
        .order_by("-created_at")
        .first()
    )


def record_consent(profile_id: UUID, action: str, items: Iterable[str]) -> ConsentRecord:
    """Tambahkan satu peristiwa persetujuan untuk versi kebijakan yang berlaku.

    Waktunya dipaksa selalu naik: dua peristiwa berurutan yang kebetulan
    tercatat pada mikrodetik yang sama akan membuat "baris terbaru" ambigu.
    """
    now = timezone.now()
    latest = latest_record(profile_id)
    if latest is not None and now <= latest.created_at:
        now = latest.created_at + timedelta(microseconds=1)
    return ConsentRecord.objects.create(
        profile_id=profile_id,
        action=action,
        policy_version=PRIVACY_POLICY_VERSION,
        items=sorted(set(items)),
        created_at=now,
    )


def current_consent(profile_id: UUID) -> ConsentRecord | None:
    """Persetujuan yang berlaku sekarang, atau None.

    None berarti belum pernah memberi, sudah menarik, atau yang diberikan
    untuk versi kebijakan lama. Ketiganya sama akibatnya: pemrosesan baru
    tidak boleh berjalan sampai pengguna menyetujui versi yang berlaku.
    """
    record = latest_record(profile_id)
    if (
        record is None
        or record.action == ConsentAction.WITHDRAWN
        or record.policy_version != PRIVACY_POLICY_VERSION
    ):
        return None
    return record


def allows_external_ai(profile_id: UUID) -> bool:
    """Apakah teks jawaban boleh dikirim ke penyedia analisis di luar negeri.

    Pasal 56 ayat (4): bila kesetaraan pelindungan di negara tujuan belum
    dapat dipastikan, transfer wajib atas persetujuan Subjek Data Pribadi.
    """
    record = current_consent(profile_id)
    return record is not None and ITEM_EXTERNAL_AI in record.items


def consent_status(profile_id: UUID) -> dict:
    """Ringkasan untuk layar Pengaturan, beserta rekam jejaknya (Pasal 32)."""
    history = list(
        ConsentRecord.objects.filter(profile_id=profile_id).order_by("-created_at")[:50]
    )
    latest = history[0] if history else None
    current = current_consent(profile_id)
    if current is not None:
        state = "given"
    elif latest is None:
        state = "none"
    elif latest.action == ConsentAction.WITHDRAWN:
        state = "withdrawn"
    else:
        state = "outdated"

    given_at = None
    if current is not None:
        given = next(
            (
                record
                for record in history
                if record.action == ConsentAction.GIVEN
                and record.policy_version == PRIVACY_POLICY_VERSION
            ),
            None,
        )
        given_at = given.created_at if given else current.created_at

    return {
        "current_version": PRIVACY_POLICY_VERSION,
        "status": state,
        "items": list(current.items) if current else [],
        "external_ai": current is not None and ITEM_EXTERNAL_AI in current.items,
        "given_at": given_at,
        "history": [
            {
                "action": record.action,
                "policy_version": record.policy_version,
                "items": list(record.items),
                "created_at": record.created_at,
            }
            for record in history
        ],
    }
